### PR TITLE
Scrub out the sensitive information from `autoregister.properties` file after reading it.

### PR DESCRIPTION
**Pull request by ketan**
The current implementation removes the `autoregister.properties` file
after reading it. Since we're going to be piggy-backing on the auto
registration mechanism to identify additional properties to associate
an agent with the specific plugin that owns it, we would like to keep
the file around across agent restarts.

### CODE REVIEW

File: agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        AutoregistrationProperties properties = new AutoregistrationProperties(configFile);
+        properties.scrubRegistrationKey();
+
+        String newContents = "" +
+                "#\n" +
+                "# file autogenerated by chef, any changes will be lost\n" +
+                "#\n" +
+                "# the registration key\n" +
+                "# The autoregister key has been intentionally replaced by Go as a security measure.\n" +
+                "agent.auto.register.key = ***** Replaced by Go *****\n" +
```
Comment (arvindsv) [inline agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java]: If the key itself is not removed or commented, won't it use `***** Replaced by Go *****` as the key next time?

Comment (ketan) [inline agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java]: Hmm, doesn't seem like there's a way to comment out a property :(

Comment (arvindsv) [inline agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java]: If an agent is already registered, does that key matter any more, any way? The server does not need to use it at all. The correct certificates have been put into the correct trust stores.

**Comment by srinivasupadhya** [comment]
you don't need to scrub since the file gets deleted after first read?

**Comment by arvindsv** [comment]
Some conversation related to this is [here](https://gitter.im/gocd/gocd?at=55a7bc83e39909763178959e). It is also related to [this](https://github.com/gocd/gocd/issues/1082#issuecomment-120886348). So, we need to find some way.

**Comment by ketan** [comment]
@arvindsv - try now? I've implemented a (low level) filter that'll intercept what's being written and comment it out if needed.

File: agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java (hunk 1 above)
Comment (ketan) [inline agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java]: The call to `register` is guarded by [`registerIfNecessary`](https://github.com/gocd/gocd/blob/b0dbaf0acfc8d4411c21773e8da56b7aefe88fe5/agent/src/com/thoughtworks/go/agent/service/SslInfrastructureService.java#L89-99).

The current implementation assumes registered implies that the agent has a GUID and the keystore has the correct certs.

So to answer your question, there is no concept of "re-register" or "un-register" that an agent can possibly initiate.

**Comment by arvindsv** [comment]
Looks good to me. I can merge it. However, how do we handle the case @srinivasupadhya brought up in the [earlier comment](https://github.com/gocd/gocd/pull/1359#issuecomment-122139597).

**Comment by ketan** [comment]
I'm not suggesting that the agent re-registers. Did you see my comment here. https://github.com/gocd/gocd/pull/1359#discussion_r34869456

**Comment by arvindsv** [comment]
Oops. Hadn't seen that. Ok, so in that case, the autoregister file will have resources, etc. which are unused beyond the first time? Could be a little misleading. But, we have to keep it around, for us to keep reusing some plugin-related settings. Right?

**Comment by ketan** [comment]
Yes, that's correct. One of the properties (`agent.elastic-agent.pluginId`) is a runtime property, the others are used only during registration.

**Comment by ketan** [comment]
We could probably scrub out the other properties if needed, and to avoid future confusion.

**Pull request merged**